teapop/man/teapop.8.in
.\" $Id: //depot/Teapop/0.3/man/teapop.8.in#7 $
.\" Copyright (c) 1999-2003 ToonTown Consulting
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the company nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\"  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R
.fi
..
.TH TEAPOP 8 @DATE@
.SH NAME
teapop \(em
a POP3 server daemon
.SH SYNOPSIS
.BI teapop
.RB [ \-dDhiLnNsSuv ]
.RB [ "\-e \fIage\fP" ]
.RB [ "\-l \fIminutes\fP" ]
.RB [ "\-p \fIhostname\fP" ]
.RB [ "\-P \fIport\fP" ]
.RB [ "\-t \fIseconds\fP" ]
.SH DESCRIPTION
.LP
.BI Teapop
is a pop3 mail program that allows remote clients to access mail.
.PP
.BI Teapop's
main goal is to be as flexible as possible, but yet be secure and fast. The
way virtual domains can be handled is somewhat unique for POP3-servers. You
no longer have to choose one way to handle all your domains, rather you can
configure how each domain will be handled separately.
.SH OPTIONS
.LP
.TP
.I "\fB\-d"
Delete messages the user has read with the RETR command.
.TP
.I "\fB\-D"
Delete messages the user has read with either the RETR or the TOP command.
Only messages that have been fully downloaded with the TOP command will be
deleted.
.TP
.I "\fB\-e\fP age"
This will force messages older than
.BI age
to be removed, after a successful exit. Age can be a mixtures of the
suffix units d(ays), w(eeks), m(onths) and y(ears). The value should
always be before the suffix and don't use any space between values. For
instance, the value 2w1m would delete all messages older then one month
and two weeks, or 44 days.

.BI NOTE:
A message which is older then the expire age will be availble during the
session, so no messages will be lost for the end user.
.TP
.I "\fB\-h"
Syntax help.
.TP
.I "\fB\-i"
Ignore the first message if it's UW IMAP's control message.
.TP
.I "\fB\-l\fP minutes"
If a dotlock file exists that is older then the specified amount of
minutes, it will be considered stale and removed. No attempt to try
to find if the process that created the file is still running will be
done. Be sure to specify an amount of minutes large enough to avoid
valid dotlock files to be removed.
.TP
.I "\fB\-L"
Activate the softlock function which will return an empty mailbox to
the client, instead of an error message, when the mailbox is locked.
.TP
.I "\fB\-n"
Don't do any kind of DNS-lookups.
.BI Teapop
normally try to verify the hostname from where the client is coming. This is
however not wanted in all environments and the saved CPU-cycle is preferred.
.TP
.I "\fB-N"
Don't resolve hostnames in teapop.passwd. If you use this switch you must
specify the IP-address, but save CPU-cycles when you run
.BI Teapop
instead.
.TP
.I "\fB\-p\fP hostname"
Directly after a user successfully authenticates, contact a DRAC server and
report the users IP address. This will be done directly after a successful
authentication and before opening the mailbox, so the IP can be reported even
if the mailbox would be locked.

Unless the flag
.B \-\-with-drac
has been specified at configure time, this switch will be silently ignored.

For more information about Dynamic Relay Authorization Control see:

  http://mail.cc.umanitoba.ca/drac/index.html
.TP
.I "\fB\-P\fP port"
In standalone mode, listen on the specified port instead of the default 110.
.TP
.I "\fB\-s"
Starts
.BI Teapop
in the background and makes it listen to port 110. Mostly known as standalone
mode.
.TP
.I "\fB\-S"
Compability mode for running over SSL tunnel. Will force
.BI Teapop
to identify as TeapopSSL to clients and log messages to syslog as
teapop_ssl.
.TP
.I "\fB\-t\fP seconds"
Changes the default timeout from 900 seconds to the number of seconds
specified. This does only affect the time waiting for a command to be sent,
and does not interfere with long downloads for instance. RFC1939 explicity
states that this MUST be set to a value of 600 (10 mins) or higher.
.TP
.I "\fB\-u"
Makes
.BI Teapop
look for X-UIDL: headers in the mail and if found use it for the UIDL
reported to the client.

.BI NOTE:
This is not for the faint hearted. If you want to use this to save CPU-cycles,
make sure you set your MTA to remove X-UIDL: headers on all incoming mail and
then add a unique X-UIDL: header. If you don't know what all this mean, this
option is NOT for you.
.TP
.I "\fB\-v"
Show version.
.SH FILES
@SYSCONFDIR@/teapop.passwd
holds information on where teapop can find the password for different
users/domains
.SH AUTHOR
Much time has been put into trying to make
.BI teapop
the most flexible POP3-server available.
.PP
Main Author(s):
.PP
.Vb 1
\&	Ibrahim 'Teaspoon' Khalifa
\&	Magnus '__ms' Stahre
.PP
Some documentation by:
.PP
.Vb 1
\&	Kitty 'Meow-Meow' Morgan
.PP
.SH SEE ALSO
inetd(8),
syslogd(8),
RFC1939
.SH BUGS
When running with the dotlock-method of locking a users mailbox,
.BI teapop
will drop root privs before creating the lockfile. Therefore the user must have
write access to the directory the user's mailbox is in. If this isn't feasable,
use flock()-method instead.
.PP
Neither mbox or POP3 handles multiple concurrent sessions well, a worst case
scenario could actually involve corrupt mailboxes.
.BI Teapop
does its' utmost to lock the file from other processes to avoid any problems,
but to ensure that no unwanted problems occurs, please make sure you use the
same locking method in both
.BI teapop
and your MTA (sendmail/qmail/postfix/etc).
.SH SECURITY
.BI Teapop
drops root privs after a valid user/pass or apop command



teapop/teapop/teapop.c
/* $Id: //depot/Teapop/0.3/teapop/teapop.c#8 $ */

/*
 * Copyright (c) 1999-2003 ToonTown Consulting
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the company nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>

#include <errno.h>
#include <setjmp.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>

#include "config.h"
#include "teapop.h"
#include "version.h"
#include "pop_auth.h"
#include "pop_dele.h"
#include "pop_file.h"
#include "pop_lock.h"
#include "pop_parse.h"
#include "pop_passwd.h"
#include "pop_signal.h"
#include "pop_socket.h"
#include "pop_stat.h"
#include "pop_strings.h"

#ifdef PROFIL
void etext __P((int));
void s_monitor __P((int));
#endif

int teapop __P((POP_INFO *));
int teapop_s __P((POP_INFO *));
void usage __P((void));

volatile int sigterm_seen = 0;

int
main(argc, argv)
int argc;
char *argv[];
{
	extern char *optarg;
	int ch, standalone, retval;
	char syslogname[20];

	POP_INFO pinfo;

	argv[0] = "teapop";
	memset(&pinfo, 0, sizeof(pinfo));
	standalone = 0;		/* runs from inetd if there is no -s */
	pinfo.ignoreimap = 0;	/* Don't ignore UW IMAP's message by default */
	pinfo.timeout = 900;	/* Disconnect an inactive user after 15 mins */
	pinfo.locktimeout = 0;	/* Don't remove dotlock files by default */

#ifdef PROFIL
	monstartup(0, etext);
	moncontrol(1);
	signal(SIGUSR1, s_monitor);
#endif

	strcpy(syslogname, "teapop");
	tzset();

	while ((ch = getopt(argc, argv, "dDe:hil:LnNp:P:sSt:vu")) != -1)
		switch (ch) {
		case 'd':
			pinfo.autodelete = 1;
			break;
		case 'D':
			pinfo.autodelete = 2;
			break;
		case 'e':
			pinfo.expire = (int)pop_string_dtot(optarg);
			break;
		case 'i':
			pinfo.ignoreimap = 1;
			break;
		case 'l':
			pinfo.locktimeout = (int) (atoi(optarg)*60);
			break;
		case 'L':
			pinfo.softlock = 1;
			break;
		case 'n':
			pinfo.nodns = 2;
			break;
		case 'N':
			if (pinfo.nodns != 2)
				pinfo.nodns = 1;
			break;
		case 'p':
			strncpy(pinfo.drachost, optarg,
			    sizeof(pinfo.drachost));
			pinfo.drachost[sizeof(pinfo.drachost) - 1] = '\0';
			break;
		case 'P':
			pinfo.localport = (unsigned short) atoi(optarg);
			break;
		case 's':
			standalone = 1;
			break;
		case 'S':
			pinfo.ssl = 1;
			strcpy(syslogname, "teapop_ssl");
			break;
		case 't':
			pinfo.timeout = (int) atoi(optarg);
			break;
		case 'v':
			version();
			/* NOT REACHED */
		case 'u':
			pinfo.useuidl = 1;
			break;
		case 'h':
		default:
			usage();
			/* NOT REACHED */
		}

	openlog(syslogname, LOG_PID | LOG_NDELAY, LOG_LOCAL0);
	pop_read_pwdinfo(&pinfo);

	pinfo.insck = fileno(stdin);
	pinfo.outsck = fileno(stderr);

	if (standalone == 1)
		retval = teapop_s(&pinfo);
	else
		retval = teapop(&pinfo);

	closelog();

	return (retval);
}

int
teapop(pinfo)
POP_INFO *pinfo;
{
	int retval;

	signal(SIGPIPE, pop_signal_sigpipe);
	signal(SIGALRM, pop_signal_sigalrm);

	/* Make sure stdin is a socket */
	if ((retval = pop_socket_init(pinfo)) != 0) {
		if (retval == 2) {
			fprintf(stderr,"connection refused by tcp_wrappers\r\n");
		} else {
			fprintf(stderr, "stdin is not a socket\r\n");
			fprintf(stderr, "If you want to run teapop in standalone " \
			    "mode you have to use the -s argument.\r\n");
		}
		return (retval);
	}

	switch (pop_auth(pinfo)) {
	case 0:		/* USER IS AUTH'ED */
		if ((retval = pop_lock_maildrop(pinfo, 1)) != 0) {
			if (retval == 3) {
				pop_socket_send(pinfo->out, "%s ehum?",
				    POP_ERR);
				pop_unlock_maildrop();
				break;
			} else if (pinfo->softlock != 1) {
				pop_socket_send(pinfo->out, "%s %s", POP_ERR,
				    POP_LOCK_ERR);
				break;
			}
			/*
			 * If we are using softlock, make sure mbox is
			 * NULL, to trick Teapop to believe the mailbox
			 * is empty.
			 */
			pinfo->mbox = NULL;
		}
		pop_get_status(pinfo);
		pop_socket_send(pinfo->out, "%s %s", POP_OK, POP_READY);
		pop_parse_cmds(pinfo);
		if ((pop_update(pinfo)) == 0)
			pop_socket_send(pinfo->out, "%s %s", POP_OK,
			    POP_QUIT_OK);
		else
			pop_socket_send(pinfo->out, "%s %s", POP_ERR,
			    POP_QUIT_ERR);
		pop_unlock_maildrop();
		break;
	case 1:		/* USER SENT QUIT IN AUTH-MODE */
		pop_socket_send(pinfo->out, "%s %s", POP_OK, POP_QUIT_AUTH);
		break;
	case 2:		/* MAX TRIES REACHED */
		pop_socket_send(pinfo->out, "%s %s", POP_OK, POP_QUIT_MAXTRY);
		break;
	default:		/* NEVER REACHED */
		break;
	}

	return (0);
}

int
teapop_s(pinfo)
POP_INFO *pinfo;
{
	int pid, sckfd, acptfd;

	if ((pid = fork()) < 0) {
		syslog(LOG_ERR, "Problem when doing fork() : %s",
		    strerror(errno));
		perror("fork");
		return (3);
	}
	setsid();

	if (pid > 0)
		exit(0);	/* exit main process */

	logpid();

	if ((sckfd = pop_socket_bind(pinfo->localport)) == 0)
		return (1);

	signal(SIGCHLD, pop_signal_sigchld);
	signal(SIGTERM, pop_signal_sigterm);

	for (;(sigterm_seen == 0);) {
		if (pop_socket_wait(sckfd) < 1)
			continue;
		if ((acptfd = accept(sckfd, NULL, NULL)) < 0)
			continue;

		switch (pid = fork()) {
		case -1:	/* fork error */
			syslog(LOG_ERR, "Problem when doing fork() : %s",
			    strerror(errno));
			return (3);
			/* NOT REACHED */
		case 0:	/* child */
			setsid();
			(void) dup2(acptfd, 0);
			(void) dup2(acptfd, 1);
			(void) dup2(acptfd, 2);
			close(sckfd);
			teapop(pinfo);
			close(acptfd);
			close(0);
			close(1);
			close(2);
			return (0);
			/* NOT REACHED */
		default:	/* parent */
			close(acptfd);
		}
	}

	/* clean up - closelog done in main() after return */
	close(sckfd);
	return (0);
}

void
usage(void)
{

	printf("Syntax: teapop [-dDhinNsSuv] [-e age] [ -l minutes ] "
	    "[-p hostname] [-P port] [-t seconds]\n");

	exit(0);
}
